src/render-to-layer.js

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import Events from './utils/events';
+import Dom from './utils/dom';
+import debounce from 'lodash.debounce';
+
+// heavily inspired by https://github.com/Khan/react-components/blob/master/js/layered-component-mixin.jsx
+const RenderToLayer = React.createClass({
+
+  componentDidMount() {
+    this._renderLayer();
+  },
+
+  componentDidUpdate() {
+    this._renderLayer();
+  },
+
+  componentWillUnmount() {
+    this._unbindClickAway();
+    if (this._layer) {
+      this._unrenderLayer();
+    }
+  },
+
+  _checkClickAway(e) {
+    if (!this.canClickAway) {
+      return;
+    }
+    const el = this._layer;
+    if (e.target !== el &&
+        !Dom.isDescendant(el, e.target) &&
+        document.documentElement.contains(e.target)) {
+      if (this.props.componentClickAway) {
+        this.props.componentClickAway(e);
+      }
+    }
+  },
+
+  _preventClickAway(e) {
+    if (e.detail === this) {
+      return;
+    }
+    this.canClickAway = false;
+  },
+
+  _allowClickAway() {
+    this.canClickAway = true;
+  },
+
+  getLayer() {
+    return this._layer;
+  },
+  
+  render() {
+    return null;
+  },
+
+  _renderLayer() {
+    if (this.props.open) {
+      if (!this._layer) {
+        this._layer = document.createElement('div');
+        document.body.appendChild(this._layer);
+      }
+      this._bindClickAway();
+      if (this.reactUnmount) {
+        this.reactUnmount.cancel();
+      }
+    } else if (this._layer) {
+      this._unbindClickAway();
+      this._unrenderLayer();
+    } else {
+      return;
+    }
+
+    // By calling this method in componentDidMount() and
+    // componentDidUpdate(), you're effectively creating a "wormhole" that
+    // funnels React's hierarchical updates through to a DOM node on an
+    // entirely different part of the page.
+
+    const layerElement = this.props.render();
+    // Renders can return null, but React.render() doesn't like being asked
+    // to render null. If we get null back from renderLayer(), just render
+    // a noscript element, like React does when an element's render returns
+    // null.
+    if (layerElement === null) {
+        this.layerElement = ReactDOM.unstable_renderSubtreeIntoContainer (this, <noscript />, this._layer);
+    } else {
+        this.layerElement = ReactDOM.unstable_renderSubtreeIntoContainer(this, layerElement, this._layer);
+    }
+
+    if (this.layerDidMount) {
```

Comment:
Hangover from it being a mixin, should probably be an optional callback on props